lib/internal/Magento/Framework/Search/Dynamic/Algorithm.php

```diff
@@ -297,9 +297,9 @@ protected function _findValueSeparator($quantileNumber, IntervalInterface $inter
                 $interval->load($intervalValuesCount + 1, $offset, $lowerValue, $this->_upperLimit)
             );
         }
-        $lastValue = $values[$intervalValuesCount - 1];
+        $lastValue = end($values);
```

Comment:
This has side-effects, unlike the original array read: you're changing the array's internal pointer.
You should `reset($values)` after getting the value.

But if you're to add 2 lines, why not keep the original code?

One more reason to not do it is that `end()` vs "reading" has a slight change of behaviour, as the former will always work; unlike `$values[$intervalValuesCount - 1]` which can yield a notice. And although that sounds good, it can also mean that, if there's a bug somewhere, `end()` won't trigger an "index not found" error.

We should stick with the original code here.